plugins/out_stackdriver/stackdriver.c

```diff
@@ -278,6 +278,241 @@ static char *get_google_token(struct flb_stackdriver *ctx)
+        v = map.ptr[i].val;
+
+        if (k.type != MSGPACK_OBJECT_STR) {
+            continue;
+        }
+
+        if (k.via.str.size == key_size && 
+            strncmp(key, (char *) k.via.str.ptr, k.via.str.size) == 0) {
+            /* make sure to free it after use */ 
+            ptr =  flb_sds_create_len(v.via.str.ptr, v.via.str.size);
+            break;
+        }
+    }
+
+    return ptr;
+}
+
+/*
+ * Given a local_resource_id, split the content using the proper separator generating
+ * a linked list to store the spliited string
+ */
+static struct mk_list *parse_local_resource_id_to_list(char *str, char *type)
+{
+    int ret = -1;
+    int max_split = -1;
+    struct mk_list *list;
+
+    /* Allocate list head */
+    list = flb_malloc(sizeof(struct mk_list));
+    if (!list) {
+        flb_errno();
+        return NULL;
+    }
+    mk_list_init(list);
+
+    /* Determinate the max split value based on type */
+    if (strncmp(type, "k8s_container", LEN_K8S_CONTAINER) == 0) {
+        /* including the prefix of tag */
+        max_split = 4;
+    }
+    else if (strncmp(type, "k8s_node", LEN_K8S_NODE) == 0) {
+        max_split = 2;
+    }
+    else if (strncmp(type, "k8s_pod", LEN_K8S_POD) == 0) {
+        max_split = 3;
+    }
+
+    /* The local_resource_id is splitted by '.' */
+    ret = flb_slist_split_string(list, str, '.', max_split);
+    if (ret == -1) {
+        flb_error("error parsing local_resource_id for type %s", type);
+        flb_free(list);
+        return NULL;
+    }
+
+    return list;
+}
+
+/* 
+*    process_local_resource_id():
+*  - extract the value from "logging.googleapis.com/local_resource_id" field
+*  - use extracted value to assign the label keys for differnet resource types
+*    that specified in the configuration of stackdriver_out plugin
+*/
+static int process_local_resource_id(const void *data, size_t bytes,
+                                     struct flb_stackdriver *ctx, char *type)
+{
+    int ret = -1;
+    int first = 1;
+    size_t off = 0;
+    msgpack_object root;
+    msgpack_object_map map;
+    msgpack_unpacked result;
+    flb_sds_t local_resource_id;
+    struct local_resource_id_list *ptr;
+    struct mk_list *list = NULL;
+    struct mk_list *tmp;
+    struct mk_list *head;
+
+    msgpack_unpacked_init(&result);
+    if (msgpack_unpack_next(&result, data, bytes, &off) == MSGPACK_UNPACK_SUCCESS) {
+        root = result.data;
+
+        if (!validate_msgpack_unpacked_data(root)) {
+            msgpack_unpacked_destroy(&result);
+            flb_plg_warn(ctx->ins, "unexpected record format");
+            return -1;
+        }
+
+        map = root.via.array.ptr[1].via.map;
+        local_resource_id = get_str_value_from_msgpack_map(map, LOCAL_RESOURCE_ID_KEY,
+                                                           LEN_LOCAL_RESOURCE_ID_KEY);
+        if (!local_resource_id) {
+            msgpack_unpacked_destroy(&result);
+            return -1;
```

Comment:
do we need log the error/warning here?

Reply:
when process_local_resource_id returns, the caller function will catch the error and log the error/warning. WDYT?